File: AccuPay/Repositories/ProductRepository.vb
Option Strict On

Imports System.Threading.Tasks
Imports AccuPay.Entity
Imports AccuPay.Enums
Imports Microsoft.EntityFrameworkCore

Namespace Global.AccuPay.Repository

    Public Class ProductRepository

        Public Async Function GetBonusTypes() _
            As Task(Of IEnumerable(Of Product))

            Dim categoryName = ProductConstant.BONUS_TYPE_CATEGORY

            Dim category = Await GetOrCreateCategoryByName(categoryName)
            Return Await GetProductsByCategory(category.RowID)

        End Function
        Public Async Function GetAllowanceTypes() _
            As Task(Of IEnumerable(Of Product))

            Dim categoryName = ProductConstant.ALLOWANCE_TYPE_CATEGORY

            Dim category = Await GetOrCreateCategoryByName(categoryName)
            Return Await GetProductsByCategory(category.RowID)

        End Function

        Public Async Function GetLeaveTypes() _
            As Task(Of IEnumerable(Of Product))

            Dim categoryName = ProductConstant.LEAVE_TYPE_CATEGORY

            Dim category = Await GetOrCreateCategoryByName(categoryName)
            Return Await GetProductsByCategory(category.RowID)

        End Function

        Public Async Function GetLoanTypes() _
            As Task(Of IEnumerable(Of Product))

            Using context As New PayrollContext

                Return Await (Await GetLoanTypesBaseQuery(context)).ToListAsync

            End Using

        End Function

        Public Async Function GetGovernmentLoanTypes() _
            As Task(Of IEnumerable(Of Product))

            Dim governmentLoans = {ProductConstant.PAG_IBIG_LOAN, ProductConstant.SSS_LOAN}

            Using context As New PayrollContext

                Return Await (Await GetLoanTypesBaseQuery(context)).
                            Where(Function(p) governmentLoans.Contains(p.PartNo)).
                            ToListAsync

            End Using

        End Function

        Public Async Function GetAdjustmentTypes() _
            As Task(Of IEnumerable(Of Product))

            Using context As New PayrollContext

                Return Await (Await GetAdjustmentTypesBaseQuery(context)).ToListAsync

            End Using

        End Function

        Public Async Function GetDeductionAdjustmentTypes() _
            As Task(Of IEnumerable(Of Product))

            Using context As New PayrollContext

                Return Await (Await GetAdjustmentTypesBaseQuery(context)).
                            Where(Function(p) p.Description = ProductConstant.ADJUSTMENT_TYPE_DEDUCTION).
                            ToListAsync

            End Using

        End Function

        Public Async Function GetAdditionAdjustmentTypes() _
            As Task(Of IEnumerable(Of Product))

            Using context As New PayrollContext

                Return Await (Await GetAdjustmentTypesBaseQuery(context)).
                            Where(Function(p) p.Description = ProductConstant.ADJUSTMENT_TYPE_ADDITION).
                            ToListAsync

            End Using

        End Function

        Public Async Function GetOrCreateLoanType(loanTypeName As String) As Task(Of Product)

            Using context = New PayrollContext()

                Dim loanType = Await context.Products.
                                Where(Function(p) Nullable.Equals(p.OrganizationID, z_OrganizationID)).
                                Where(Function(p) p.PartNo.ToLower = loanTypeName.ToLower).
                                FirstOrDefaultAsync

                If loanType Is Nothing Then
                    loanType = Await AddLoanType(loanTypeName)
                End If

                Return loanType

            End Using

        End Function

        Public Async Function GetOrCreateAllowanceType(allowanceTypeName As String) As Task(Of Product)

            Using context = New PayrollContext()

                Dim allowanceType = Await context.Products.
                                    Where(Function(p) Nullable.Equals(p.OrganizationID, z_OrganizationID)).
                                    Where(Function(p) p.PartNo.ToLower = allowanceTypeName.ToLower).
                                    FirstOrDefaultAsync

                If allowanceType Is Nothing Then
                    allowanceType = Await AddAllowanceType(allowanceTypeName)
                End If

                Return allowanceType

            End Using

        End Function

        Public Async Function GetOrCreateAdjustmentType(adjustmentTypeName As String) As Task(Of Product)

            Using context = New PayrollContext()

                Dim adjustmentType = Await context.Products.
                                    Where(Function(p) Nullable.Equals(p.OrganizationID, z_OrganizationID)).
                                    Where(Function(p) p.PartNo.ToLower = adjustmentTypeName.ToLower).
                                    FirstOrDefaultAsync

                If adjustmentType Is Nothing Then
                    adjustmentType = Await AddAdjustmentType(adjustmentTypeName)
                End If

                Return adjustmentType

            End Using

        End Function

        Public Async Function GetOrCreateBonusType(bonusTypeName As String) As Task(Of Product)

            Using context = New PayrollContext()

                Dim bonusType = Await context.Products.
                                Where(Function(p) Nullable.Equals(p.OrganizationID, z_OrganizationID)).
                                Where(Function(p) p.PartNo.ToLower = bonusTypeName.ToLower).
                                FirstOrDefaultAsync

                If bonusType Is Nothing Then
                    bonusType = Await AddBonusType(bonusTypeName)
                End If

                Return bonusType

            End Using

        End Function

        Public Async Function AddBonusType(loanName As String, Optional isTaxable As Boolean = False, Optional throwError As Boolean = True) _
            As Task(Of Product)

            Dim product As New Product

            product.Category = ProductConstant.BONUS_TYPE_CATEGORY

            Return Await AddProduct(loanName, throwError, product, isTaxable)

        End Function

        Public Async Function AddLoanType(loanName As String, Optional throwError As Boolean = True) _
            As Task(Of Product)

            Dim product As New Product

            product.Category = ProductConstant.LOAN_TYPE_CATEGORY

            Return Await AddProduct(loanName, throwError, product)

        End Function

        Public Async Function AddAllowanceType(allowanceName As String, Optional throwError As Boolean = True) _
            As Task(Of Product)

            Dim product As New Product

            product.Category = ProductConstant.ALLOWANCE_TYPE_CATEGORY

            Return Await AddProduct(allowanceName, throwError, product)

        End Function

        Public Async Function AddAdjustmentType(
                                    adjustmentName As String,
                                    Optional adjustmentType As AdjustmentType = AdjustmentType.Blank,
                                    Optional comments As String = "",
                                    Optional throwError As Boolean = True) _
            As Task(Of Product)

            Dim product As New Product

            product.Comments = comments
            product.Description = DetermineAdjustmentTypeString(adjustmentType)

            product.Category = ProductConstant.ADJUSTMENT_TYPE_CATEGORY

            Return Await AddProduct(adjustmentName, throwError, product)

        End Function

        Public Async Function Delete(id As Integer, Optional throwError As Boolean = True) _
            As Task(Of Boolean)

            Using context = New PayrollContext()

                Dim product = Await context.Products.FirstOrDefaultAsync(Function(p) p.RowID.Value = id)

                If product Is Nothing Then
                    If throwError Then
                        Throw New ArgumentException("The data that you want to delete is already gone. Please reopen the form to refresh the page.")
                    Else
                        Return False
                    End If

                End If

                context.Products.Remove(product)

                Await context.SaveChangesAsync()

                Return True
            End Using

        End Function

        Public Async Function UpdateAdjustmentType(id As Integer, adjustmentName As String, code As String, Optional throwError As Boolean = True) _
            As Task(Of Product)

            Using context = New PayrollContext()

                Dim product = Await context.Products.FirstOrDefaultAsync(Function(p) p.RowID.Value = id)

                If product Is Nothing AndAlso throwError Then
                    Throw New ArgumentException("There was a problem in updating the adjustment type. Please reopen the form and try again.")
                End If

                product.PartNo = adjustmentName.Trim()
                product.Name = adjustmentName.Trim()

                product.Comments = code
                product.LastUpdBy = z_User

                Await context.SaveChangesAsync()

                Dim newProduct = Await context.Products.
                    FirstOrDefaultAsync(Function(p) Nullable.Equals(p.RowID, product.RowID))

                If newProduct Is Nothing AndAlso throwError Then
                    Throw New ArgumentException("There was a problem inserting the new adjustment type. Please try again.")
                End If

                Return newProduct
            End Using

        End Function

        Public Async Function CheckIfProductExists(productName As String, categoryId As Integer?) As Task(Of Boolean)

            Using context As New PayrollContext

                Return Await context.Products.
                                Where(Function(p) p.PartNo.Trim = productName.Trim).
                                Where(Function(p) p.CategoryID.Value = categoryId.Value).
                                Where(Function(p) p.OrganizationID.Value = z_OrganizationID).
                                AnyAsync
            End Using

        End Function

        Public Function ConvertToStringList(products As IEnumerable(Of Product), Optional columnName As String = "PartNo") _
            As List(Of String)

            Dim stringList As List(Of String)
            stringList = New List(Of String)

            For Each product In products

                Select Case columnName
                    Case "Name"
                        stringList.Add(product.PartNo)

                    Case Else
                        stringList.Add(product.Name)
                End Select
            Next

            Return stringList.OrderBy(Function(s) s).ToList

        End Function

#Region "Private Functions"

        'this may only apply to "Loan Type" use with caution
        Private Shared Async Function GetOrCreateCategoryByName(categoryName As String) As Task(Of Category)
            Using context = New PayrollContext()

                Dim categoryProduct = Await context.Categories.
                                    Where(Function(c) Nullable.Equals(c.OrganizationID, z_OrganizationID)).
                                    Where(Function(c) c.CategoryName = categoryName).
                                    FirstOrDefaultAsync

                If categoryProduct Is Nothing Then
                    'get the existing category with same name to use as CategoryID
                    Dim existingCategoryProduct = Await context.Categories.
                                    Where(Function(c) c.CategoryName = categoryName).
                                    FirstOrDefaultAsync

                    Dim existingCategoryProductId = existingCategoryProduct?.RowID

                    categoryProduct = New Category
                    categoryProduct.CategoryID = existingCategoryProductId
                    categoryProduct.CategoryName = categoryName
                    categoryProduct.OrganizationID = z_OrganizationID
                    categoryProduct.CatalogID = Nothing
                    categoryProduct.LastUpd = Date.Now

                    context.Categories.Add(categoryProduct)
                    context.SaveChanges()

                    'if there is no existing category with same name,
                    'use the newly added category's RowID as its CategoryID

                    If existingCategoryProductId Is Nothing Then

                        Try
                            categoryProduct.CategoryID = categoryProduct.RowID
                            Await context.SaveChangesAsync()
                        Catch ex As Exception
                            'if for some reason hindi na update, we can't let that row
                            'to have no CategoryID so dapat i-delete rin yung added category
                            context.Categories.Remove(categoryProduct)
                            context.SaveChanges()

                            Throw ex
                        End Try

                    End If
                End If

                If categoryProduct Is Nothing Then
                    Dim ex = New Exception("ProductRepository->GetOrCreate: Category not found.")
                    Throw ex
                End If

                Return categoryProduct

            End Using
        End Function

        Private Async Function GetProductsByCategory(categoryId As Integer?) _
            As Task(Of IEnumerable(Of Product))

            Using context = New PayrollContext()

                Dim listOfValues = Await GetProductsByCategoryBaseQuery(categoryId, context).
                                            ToListAsync

                Return listOfValues

            End Using

        End Function

        Private Function GetProductsByCategoryBaseQuery(categoryId As Integer?, context As PayrollContext) _
            As IQueryable(Of Product)

            Return context.Products.
                                Where(Function(p) Nullable.Equals(p.OrganizationID, z_OrganizationID)).
                                Where(Function(p) Nullable.Equals(p.CategoryID, categoryId))

        End Function

        Private Async Function GetAdjustmentTypesBaseQuery(context As PayrollContext) As Task(Of IQueryable(Of Product))

            Dim categoryName = ProductConstant.ADJUSTMENT_TYPE_CATEGORY

            Dim category = Await GetOrCreateCategoryByName(categoryName)
            Return GetProductsByCategoryBaseQuery(category.RowID, context)

        End Function

        Private Async Function GetLoanTypesBaseQuery(context As PayrollContext) As Task(Of IQueryable(Of Product))

            Dim categoryName = ProductConstant.LOAN_TYPE_CATEGORY

            Dim category = Await GetOrCreateCategoryByName(categoryName)
            Return GetProductsByCategoryBaseQuery(category.RowID, context)

        End Function

        Private Async Function AddProduct(productName As String, throwError As Boolean, product As Product, Optional isTaxable As Boolean = False) As Task(Of Product)

            Dim categoryId = (Await GetOrCreateCategoryByName(product.Category))?.RowID

            If categoryId Is Nothing Then

                If throwError Then
                    Throw New ArgumentException("There was a problem on saving the data. Please try again.")
                Else
                    Return Nothing
                End If

            End If

            If Await CheckIfProductExists(productName, categoryId) Then

                If throwError Then
                    Throw New ArgumentException("Product already exists.")
                Else
                    Return Nothing
                End If

            End If

            product.CategoryID = categoryId

            product.PartNo = productName.Trim()
            product.Name = productName.Trim()
            product.Status = If(isTaxable, "1", "0")
            product.Created = Date.Now
            product.CreatedBy = z_User
            product.OrganizationID = z_OrganizationID

            Using context = New PayrollContext()

                context.Products.Add(product)

                Await context.SaveChangesAsync()

                Dim newProduct = Await context.Products.
                    FirstOrDefaultAsync(Function(p) Nullable.Equals(p.RowID, product.RowID))

                If newProduct Is Nothing AndAlso throwError Then
                    Throw New ArgumentException("There was a problem on saving the data. Please try again.")
                End If

                Return newProduct
            End Using
        End Function

        Private Shared Function DetermineAdjustmentTypeString(adjustmentType As AdjustmentType) As String
            If adjustmentType = AdjustmentType.Deduction Then

                Return ProductConstant.ADJUSTMENT_TYPE_DEDUCTION

            ElseIf adjustmentType = AdjustmentType.OtherIncome Then

                Return ProductConstant.ADJUSTMENT_TYPE_ADDITION
            Else

                Return String.Empty

            End If
        End Function

#End Region

    End Class

End Namespace

File: Accupay.Data/Repositories/ProductRepository.cs
using AccuPay.Data.Entities;
using AccuPay.Data.Enums;
using AccuPay.Data.Helpers;
using AccuPay.Utilities.Extensions;
using Microsoft.EntityFrameworkCore;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace AccuPay.Data.Repositories
{
    public class ProductRepository
    {
        private readonly PayrollContext _context;

        public ProductRepository(PayrollContext context)
        {
            _context = context;
        }

        #region Save

        public async Task DeleteAsync(int id)
        {
            var product = await _context.Products.FirstOrDefaultAsync(p => p.RowID == id);

            _context.Products.Remove(product);

            await _context.SaveChangesAsync();
        }

        public async Task<Product> AddBonusTypeAsync(string loanName, int organizationId, int userId, bool isTaxable = false)
        {
            Product product = new Product
            {
                Category = ProductConstant.BONUS_TYPE_CATEGORY
            };

            return await AddProductAsync(loanName, product, organizationId, userId, isTaxable);
        }

        public async Task<Product> AddDisciplinaryTypeAsync(string name, int organizationId, int userId, string description)
        {
            Product product = new Product
            {
                Category = ProductConstant.EMPLOYEE_DISCIPLINARY_CATEGORY,
                Description = description,
                ActiveData = true
            };

            return await AddProductAsync(name, product, organizationId, userId);
        }

        public async Task<Product> AddLeaveTypeAsync(string leaveName, int organizationId, int userId)
        {
            Product product = new Product
            {
                Category = ProductConstant.LEAVE_TYPE_CATEGORY,
            };

            return await AddProductAsync(leaveName, product, organizationId, userId);
        }

        public async Task<Product> AddAllowanceTypeAsync(string allowanceName, int organizationId, int userId)
        {
            Product product = new Product
            {
                Category = ProductConstant.ALLOWANCE_TYPE_CATEGORY,
                ActiveData = true,
                AllocateBelowSafetyFlag = '0'
            };

            return await AddProductAsync(allowanceName, product, organizationId, userId);
        }

        public async Task<Product> AddLoanTypeAsync(string loanName, int organizationId, int userId)
        {
            Product product = new Product
            {
                Category = ProductConstant.LOAN_TYPE_CATEGORY
            };

            return await AddProductAsync(loanName, product, organizationId, userId);
        }

        public async Task<List<Product>> AddManyLoanTypeAsync(List<string> loanNames, int organizationId, int userId)
        {
            List<Product> products = new List<Product>();
            foreach (var loanName in loanNames)
            {
                var product = new Product()
                {
                    Category = ProductConstant.LOAN_TYPE_CATEGORY
                };

                var addedLoanType = await AddProductAsync(loanName, product, organizationId, userId);

                products.Add(addedLoanType);
            }

            return products;
        }

        public async Task<Product> AddAdjustmentTypeAsync(string adjustmentName, int organizationId, int userId, AdjustmentType adjustmentType = AdjustmentType.Blank, string comments = "")
        {
            Product product = new Product
            {
                Comments = comments,
                Description = DetermineAdjustmentTypeString(adjustmentType),
                Category = ProductConstant.ADJUSTMENT_TYPE_CATEGORY
            };

            return await AddProductAsync(adjustmentName, product, organizationId, userId);
        }

        private async Task<Product> AddProductAsync(string productName, Product product, int organizationId, int userId, bool isTaxable = false)
        {
            var categoryId = (await GetOrCreateCategoryByName(product.Category, organizationId))?.RowID;

            if (categoryId == null)
                throw new ArgumentException("There was a problem on saving the data. Please try again.");

            if (await CheckIfProductExists(productName, categoryId.Value, organizationId))
                throw new ArgumentException("Product already exists.");

            product.CategoryID = categoryId;

            product.PartNo = productName.Trim();
            product.Name = productName.Trim();
            product.Status = isTaxable ? "1" : "0";

            product.Created = DateTime.Now;
            product.CreatedBy = userId;
            product.OrganizationID = organizationId;

            _context.Products.Add(product);

            await _context.SaveChangesAsync();

            var newProduct = await _context.Products.FirstOrDefaultAsync(p => p.RowID == product.RowID);

            return newProduct;
        }

        public async Task<Product> UpdateAdjustmentTypeAsync(int id, int userId, string adjustmentName, string code)
        {
            var product = await _context.Products.FirstOrDefaultAsync(p => p.RowID == id);

            if (product == null) return null;

            product.PartNo = adjustmentName.Trim();
            product.Name = adjustmentName.Trim();

            product.Comments = code;
            product.LastUpdBy = userId;

            await _context.SaveChangesAsync();

            var newProduct = await _context.Products.FirstOrDefaultAsync(p => p.RowID == product.RowID);

            return newProduct;
        }

        public async Task<Product> UpdateDisciplinaryTypeAsync(int id, int userId, string adjustmentName, string description)
        {
            var product = await _context.Products.FirstOrDefaultAsync(p => p.RowID == id);

            if (product == null) return null;

            product.PartNo = adjustmentName.Trim();
            product.Name = adjustmentName.Trim();
            product.Description = description;
            product.LastUpdBy = userId;

            await _context.SaveChangesAsync();

            var newProduct = await _context.Products.FirstOrDefaultAsync(p => p.RowID == product.RowID);

            return newProduct;
        }

        public async Task UpdateLoanTypeAsync(int id, string loanTypeName)
        {
            var product = await GetByIdAsync(id);

            product.PartNo = loanTypeName;
            product.Name = loanTypeName;

            _context.Entry(product).State = EntityState.Modified;

            await _context.SaveChangesAsync();
        }

        public async Task DeleteLoanTypeAsync(int id)
        {
            var product = await GetByIdAsync(id);

            _context.Entry(product).State = EntityState.Deleted;
            //_context.Products.Remove(product);

            await _context.SaveChangesAsync();
        }

        #endregion Save

        #region Queries

        #region Single entity

        public async Task<Product> GetByIdAsync(int id)
        {
            return await _context.Products.FirstOrDefaultAsync(l => l.RowID == id);
        }

        public async Task<ICollection<Product>> GetManyByIdsAsync(int[] ids)
        {
            return await _context.Products
                .AsNoTracking()
                .Where(x => ids.Contains(x.RowID.Value))
                .ToListAsync();
        }

        public async Task<Product> GetOrCreateAdjustmentTypeAsync(string adjustmentTypeName, int organizationId, int userId)
        {
            return await GetOrCreateTypeAsync(
                categoryName: ProductConstant.ADJUSTMENT_TYPE_CATEGORY,
                typeName: adjustmentTypeName,
                organizationId: organizationId,
                userId: userId);
        }

        public async Task<Product> GetOrCreateAllowanceTypeAsync(string allowanceTypeName, int organizationId, int userId)
        {
            return await GetOrCreateTypeAsync(
                categoryName: ProductConstant.ALLOWANCE_TYPE_CATEGORY,
                typeName: allowanceTypeName,
                organizationId: organizationId,
                userId: userId);
        }

        public async Task<Product> GetOrCreateLeaveTypeAsync(string leaveTypeName, int organizationId, int userId)
        {
            return await GetOrCreateTypeAsync(
                categoryName: ProductConstant.LEAVE_TYPE_CATEGORY,
                typeName: leaveTypeName,
                organizationId: organizationId,
                userId: userId);
        }

        public async Task<Product> GetOrCreateLoanTypeAsync(string loanTypeName, int organizationId, int userId)
        {
            return await GetOrCreateTypeAsync(
                categoryName: ProductConstant.LOAN_TYPE_CATEGORY,
                typeName: loanTypeName,
                organizationId: organizationId,
                userId: userId);
        }

        // TODO: maybe move this to data service because of the ArgumentException
        private async Task<Product> GetOrCreateTypeAsync(string categoryName, string typeName, int organizationId, int userId)
        {
            var categoryId = (await GetOrCreateCategoryByName(categoryName, organizationId))?.RowID;

            if (categoryId == null)
                throw new ArgumentException("There was a problem on fetching the data. Please try again.");

            var productType = await GetProductByNameAndCategory(typeName, categoryId.Value, organizationId);

            if (productType == null)
            {
                switch (categoryName)
                {
                    case ProductConstant.ADJUSTMENT_TYPE_CATEGORY:
                        productType = await AddAdjustmentTypeAsync(typeName, organizationId: organizationId, userId: userId);
                        break;

                    case ProductConstant.LEAVE_TYPE_CATEGORY:
                        productType = await AddLeaveTypeAsync(typeName, organizationId: organizationId, userId: userId);
                        break;

                    case ProductConstant.ALLOWANCE_TYPE_CATEGORY:
                        productType = await AddAllowanceTypeAsync(typeName, organizationId: organizationId, userId: userId);
                        break;

                    case ProductConstant.LOAN_TYPE_CATEGORY:
                        productType = await AddLoanTypeAsync(typeName, organizationId: organizationId, userId: userId);
                        break;
                }
            }

            return productType;
        }

        #endregion Single entity

        #region List of entities

        public async Task<ICollection<Product>> GetBonusTypesAsync(int organizationId)
        {
            var categoryName = ProductConstant.BONUS_TYPE_CATEGORY;

            var category = await GetOrCreateCategoryByName(categoryName, organizationId);
            return await GetProductsByCategory(category.RowID, organizationId);
        }

        public async Task<ICollection<Product>> GetDisciplinaryTypesAsync(int organizationId)
        {
            var categoryName = ProductConstant.EMPLOYEE_DISCIPLINARY_CATEGORY;

            var category = await GetOrCreateCategoryByName(categoryName, organizationId);
            return await GetProductsByCategory(category.RowID, organizationId);
        }

        public async Task<ICollection<Product>> GetAllowanceTypesAsync(int organizationId)
        {
            var categoryName = ProductConstant.ALLOWANCE_TYPE_CATEGORY;

            var category = await GetOrCreateCategoryByName(categoryName, organizationId);
            return await GetProductsByCategory(category.RowID, organizationId);
        }

        public async Task<ICollection<Product>> GetLeaveTypesAsync(int organizationId)
        {
            var categoryName = ProductConstant.LEAVE_TYPE_CATEGORY;

            var category = await GetOrCreateCategoryByName(categoryName, organizationId);
            return await GetProductsByCategory(category.RowID, organizationId);
        }

        public async Task<ICollection<Product>> GetLoanTypesAsync(int organizationId)
        {
            return await (await GetLoanTypesBaseQuery(organizationId))
                .ToListAsync();
        }

        public async Task<ICollection<Product>> GetGovernmentLoanTypesAsync(int organizationId)
        {
            string[] governmentLoans = {
                ProductConstant.PAG_IBIG_LOAN,
                ProductConstant.SSS_LOAN
            };

            return await (await GetLoanTypesBaseQuery(organizationId))
                    .Where(p => governmentLoans.Contains(p.PartNo))
                .ToListAsync();
        }

        public async Task<ICollection<Product>> GetAdjustmentTypesAsync(int organizationId)
        {
            return await (await GetAdjustmentTypesBaseQuery(organizationId))
                .ToListAsync();
        }

        public async Task<ICollection<Product>> GetDeductionAdjustmentTypesAsync(int organizationId)
        {
            return await (await GetAdjustmentTypesBaseQuery(organizationId))
                    .Where(p => p.Description == ProductConstant.ADJUSTMENT_TYPE_DEDUCTION)
                .ToListAsync();
        }

        public async Task<ICollection<Product>> GetAdditionAdjustmentTypesAsync(int organizationId)
        {
            return await (await GetAdjustmentTypesBaseQuery(organizationId))
                    .Where(p => p.Description == ProductConstant.ADJUSTMENT_TYPE_ADDITION)
                .ToListAsync();
        }

        public async Task<PaginatedList<Product>> GetPaginatedLoanTypeListAsync(PageOptions options, string searchTerm, int organizationId)
        {
            var query = _context.Products
                .Where(p => p.CategoryEntity.CategoryName == ProductConstant.LOAN_TYPE_CATEGORY)
                .Where(p => p.OrganizationID == organizationId)
                .AsQueryable();

            if (!string.IsNullOrWhiteSpace(searchTerm))
            {
                searchTerm = $"%{searchTerm}%";

                query = query.Where(x =>
                    EF.Functions.Like(x.Name, searchTerm) ||
                    EF.Functions.Like(x.PartNo, searchTerm));
            }

            var loanTypes = await query.Page(options).ToListAsync();
            var count = await query.CountAsync();

            return new PaginatedList<Product>(loanTypes, count);
        }

        #endregion List of entities

        #endregion Queries

        #region Others

        public List<string> ConvertToStringList(ICollection<Product> products, string columnName = "PartNo")
        {
            List<string> stringList;
            stringList = new List<string>();

            foreach (var product in products)
            {
                switch (columnName)
                {
                    case "Name":
                        stringList.Add(product.Name);
                        break;

                    default:
                        stringList.Add(product.PartNo);
                        break;
                }
            }

            return stringList.OrderBy(s => s).ToList();
        }

        #endregion Others

        #region Private helper methods

        private async Task<Product> GetProductByNameAndCategory(string productName, int categoryId, int organizationId)
        {
            return await CreateBaseQueryByCategory(categoryId, organizationId)
                .Where(p => p.PartNo.Trim().ToLower() == productName.ToTrimmedLowerCase())
                .FirstOrDefaultAsync();
        }

        private async Task<bool> CheckIfProductExists(string productName, int categoryId, int organizationId)
        {
            var product = await GetProductByNameAndCategory(
                productName,
                categoryId: categoryId,
                organizationId: organizationId);

            return product != null;
        }

        // this may only apply to "Loan Type" use with caution
        private async Task<Category> GetOrCreateCategoryByName(string categoryName, int organizationId)
        {
            var categoryProduct = await _context.Categories
                .Where(c => c.OrganizationID == organizationId)
                .Where(c => c.CategoryName.Trim().ToLower() == categoryName.ToTrimmedLowerCase())
                .FirstOrDefaultAsync();

            if (categoryProduct == null)
            {
                // get the existing category with same name to use as CategoryID
                // this is due to the design of the database wanting the category to have a CategoryID
                // to maybe group them across different organization but never used them in the application ever
                var existingCategoryProduct = await _context.Categories
                    .Where(c => c.CategoryName == categoryName)
                    .FirstOrDefaultAsync();

                var existingCategoryProductId = existingCategoryProduct?.RowID;

                categoryProduct = new Category
                {
                    CategoryID = existingCategoryProductId,
                    CategoryName = categoryName,
                    OrganizationID = organizationId,
                    CatalogID = null,
                    LastUpd = DateTime.Now
                };

                _context.Categories.Add(categoryProduct);
                await _context.SaveChangesAsync();

                // if there is no existing category with same name,
                // use the newly added category's RowID as its CategoryID

                if (existingCategoryProductId == null)
                {
                    try
                    {
                        categoryProduct.CategoryID = categoryProduct.RowID;
                        await _context.SaveChangesAsync();
                    }
                    catch (Exception ex)
                    {
                        // if for some reason hindi na update, we can't let that row
                        // to have no CategoryID so dapat i-delete rin yung added category
                        _context.Categories.Remove(categoryProduct);
                        await _context.SaveChangesAsync();

                        throw ex;
                    }
                }
            }

            if (categoryProduct == null)
            {
                var ex = new Exception("ProductRepository->GetOrCreate: Category not found.");
                throw ex;
            }

            return categoryProduct;
        }

        internal async Task<bool> CheckIfAlreadyUsedInAllowancesAsync(string allowanceType)
        {
            return await _context.AllowanceItems
                .Include(x => x.Allowance)
                .Include(x => x.Allowance.Product)
                .Where(x => x.Allowance.Product.PartNo == allowanceType)
                .AnyAsync();
        }

        private async Task<ICollection<Product>> GetProductsByCategory(int categoryId, int organizationId)
        {
            var listOfValues = await CreateBaseQueryByCategory(categoryId, organizationId)
                .ToListAsync();

            return listOfValues;
        }

        private async Task<IQueryable<Product>> GetAdjustmentTypesBaseQuery(int organizationId)
        {
            var categoryName = ProductConstant.ADJUSTMENT_TYPE_CATEGORY;

            var category = await GetOrCreateCategoryByName(categoryName, organizationId);
            return CreateBaseQueryByCategory(category.RowID, organizationId);
        }

        private async Task<IQueryable<Product>> GetLoanTypesBaseQuery(int organizationId)
        {
            var categoryName = ProductConstant.LOAN_TYPE_CATEGORY;

            var category = await GetOrCreateCategoryByName(categoryName, organizationId);
            return CreateBaseQueryByCategory(category.RowID, organizationId);
        }

        private IQueryable<Product> CreateBaseQueryByCategory(int categoryId, int organizationId)
        {
            return _context.Products
                .Where(p => p.OrganizationID == organizationId)
                .Where(p => p.CategoryID == categoryId);
        }

        private static string DetermineAdjustmentTypeString(AdjustmentType adjustmentType)
        {
            if (adjustmentType == AdjustmentType.Deduction)
                return ProductConstant.ADJUSTMENT_TYPE_DEDUCTION;
            else if (adjustmentType == AdjustmentType.OtherIncome)
                return ProductConstant.ADJUSTMENT_TYPE_ADDITION;
            else
                return string.Empty;
        }

        #endregion Private helper methods
    }
}